### DISCUSSION TITLE
Escaping Special Characters such as @ signs in Passwords

### DISCUSSION THREAD

**Discussion by ChrisDoddGit** [sqlalchemy/sqlalchemy#9827] [Usage Questions]
Hello,

When reading the docs on escaping special characters in passwords: https://docs.sqlalchemy.org/en/20/core/engines.html#escaping-special-characters-such-as-signs-in-passwords

I have some unexpected behavior. I currently authenticate using the following which works well:
```
password = 'mypassword'
uri = f'mysql+mysqldb://myuser:{password}@myhost
with create_engine(uri).connect() as engine:  # connects properly
```

However, when I change my password to include a special character and encode it as the docs suggest, It does not work. 

```
password = 'my@password'
password = urllib.parse.quote_plus(password)  # suggested by docs
uri = f'mysql+mysqldb://myuser:{password}@myhost
with create_engine(uri).connect() as engine:  # does not connect!

print(uri) -> 'mysql+mysqldb://myuser:my%40password@myhost'
```

I see it is escaping properly, unfortunately the connection is not working.

NOTE: I see we can construct programmatically to avoid the parsing: 
https://docs.sqlalchemy.org/en/20/core/engines.html#creating-urls-programmatically

However, this will not work as legacy code is constructing the uri as demonstrated above.

Thank you,
Chris

**Comment by zzzeek** [comment]
can you try the programmatic URL approach first to verify the password actually works.   

Here's a series of transformations which you can use to get the connect arguments for the DBAPI directly, and compare URL-encoded strings with URL objects:

```py
>>> import urllib
>>> password = 'my@password'
>>> password = urllib.parse.quote_plus(password) 
>>> uri = f'mysql+mysqldb://myuser:{password}@myhost'
>>> from sqlalchemy.engine import url
>>> u1 = url.URL("mysql+mysqldb", username="myuser", password="my@password", host="myhost")
>>> u2 = url.make_url(uri)
>>> u1 == u2
True

>>> from sqlalchemy import create_engine
>>> create_engine("mysql+mysqldb://").dialect.create_connect_args(u1)
[[], {'host': 'myhost', 'user': 'myuser', 'passwd': 'my@password', 'client_flag': 2}]
>>> create_engine("mysql+mysqldb://").dialect.create_connect_args(u2)
[[], {'host': 'myhost', 'user': 'myuser', 'passwd': 'my@password', 'client_flag': 2}]

```


the arguments that come from create_connect_args can be fed directly into [mysqlclient.connect()](https://mysqlclient.readthedocs.io/user_guide.html#functions-and-attributes) so you can trace the password all the way through SQLAlchemy
